File: forge-core/internal/authenticatedadrlifecyclecontract/authority.go
package authenticatedadrlifecyclecontract

import (
	"fmt"

	approvalcontract "forgeos/forge-core/internal/authenticatedadrapprovalcontract"
)

var signatureProfileConstants = map[string]string{
	"algorithm":           "Ed25519",
	"api_version":         signatureProfileAPI,
	"canonicalization":    canonicalization,
	"digest_algorithm":    "SHA-256",
	"kind":                "SignatureProfile",
	"message_preimage":    "domain_separator_utf8_nul_then_raw_32_byte_sha256_digest",
	"profile_id":          signatureProfileID,
	"public_key_encoding": "base64url_unpadded_32_bytes",
	"signature_encoding":  "base64url_unpadded_64_bytes",
}

func validateSignatureProfile(value any) (map[string]any, error) {
	fields := []string{"algorithm", "api_version", "canonicalization", "digest_algorithm",
		"kind", "message_preimage", "profile_id", "profile_sha256",
		"public_key_encoding", "signature_encoding"}
	node, err := requireKeys(value, "SignatureProfile", fields...)
	if err != nil {
		return nil, err
	}
	if _, err = boundedCanonicalJSON(node, maxProfileBytes, "SignatureProfile"); err != nil {
		return nil, err
	}
	for field, expected := range signatureProfileConstants {
		if node[field] != expected {
			return nil, fmt.Errorf("SignatureProfile.%s drifted from v1", field)
		}
	}
	digest, err := signatureProfileSHA256(node)
	if err != nil || node["profile_sha256"] != digest || digest != signatureProfileSHA256Pin {
		return nil, fmt.Errorf("SignatureProfile self digest does not match")
	}
	return node, nil
}

func validateApprovalRoot(value any) (*approvalcontract.TrustRoot, error) {
	raw, err := boundedCanonicalJSON(value, maxGoldenBytes, "approval trust root")
	if err != nil {
		return nil, err
	}
	root, err := approvalcontract.DecodeCanonicalTrustRoot(raw)
	if err != nil {
		return nil, fmt.Errorf("approval trust root: %w", err)
	}
	return root, nil
}

func validateLifecycleRoot(value any, profileHash string) (map[string]any, error) {
	label := "ArchitectureDecisionLifecycleTrustRoot"
	fields := []string{"api_version", "canonicalization", "keys", "kind", "profile_id",
		"root_sha256", "signature_profile_sha256", "trust_domain", "trust_epoch"}
	node, err := requireKeys(value, label, fields...)
	if err != nil {
		return nil, err
	}
	if _, err = boundedCanonicalJSON(node, maxRootBytes, label); err != nil {
		return nil, err
	}
	if node["api_version"] != trustRootAPI || node["canonicalization"] != canonicalization ||
		node["kind"] != label || node["profile_id"] != profileID {
		return nil, fmt.Errorf("%s envelope drifted from v1", label)
	}
	if err = validateLifecycleRootScalars(node, profileHash); err != nil {
		return nil, err
	}
	if err = validateLifecycleRootKeys(node); err != nil {
		return nil, err
	}
	digest, err := trustRootSHA256(node)
	if err != nil || node["root_sha256"] != digest {
		return nil, fmt.Errorf("%s self digest does not match", label)
	}
	return node, nil
}

func validateLifecycleRootScalars(node map[string]any, profileHash string) error {
	if _, err := textValue(node["trust_domain"], "lifecycle_root.trust_domain", 160); err != nil {
		return err
	}
	if _, err := intValue(node["trust_epoch"], "lifecycle_root.trust_epoch", 1, maxInt64); err != nil {
		return err
	}
	if node["signature_profile_sha256"] != profileHash {
		return fmt.Errorf("lifecycle trust root does not bind the SignatureProfile")
	}
	_, err := shaValue(node["root_sha256"], "lifecycle_root.root_sha256")
	return err
}

func validateLifecycleRootKeys(root map[string]any) error {
	items, err := sortedUniqueNodes(root["keys"], "lifecycle_root.keys", 2)
	if err != nil {
		return err
	}
	keys := make([]map[string]any, len(items))
	for index, item := range items {
		keys[index], err = validateLifecycleRootKey(item, index)
		if err != nil {
			return err
		}
	}
	if err = validateLifecycleUsages(keys); err != nil {
		return err
	}
	return validateLifecycleKeyDistinctness(keys, root["trust_domain"].(string))
}

func validateLifecycleRootKey(value any, index int) (map[string]any, error) {
	label := fmt.Sprintf("lifecycle_root.keys[%d]", index)
	node, err := requireKeys(value, label, "key_id", "principal", "public_key_base64url", "usage")
	if err != nil {
		return nil, err
	}
	if _, err = textValue(node["key_id"], label+".key_id", 160); err != nil {
		return nil, err
	}
	usage, err := enumValue(node["usage"], label+".usage", requestKeyUsage, stateKeyUsage)
	if err != nil {
		return nil, err
	}
	allowed := []string{"service"}
	if usage == requestKeyUsage {
		allowed = []string{"agent", "human", "operator", "service"}
	}
	if _, err = validatePrincipal(node["principal"], label+".principal", allowed...); err != nil {
		return nil, err
	}
	if _, err = fixedBase64URL(node["public_key_base64url"], label+".public_key_base64url", 32); err != nil {
		return nil, err
	}
	return node, nil
}

func validateLifecycleUsages(keys []map[string]any) error {
	counts := map[string]int{}
	for _, key := range keys {
		counts[key["usage"].(string)]++
	}
	if counts[requestKeyUsage] != 1 || counts[stateKeyUsage] != 1 {
		return fmt.Errorf("lifecycle trust root requires one key for each exact usage")
	}
	return nil
}

func validateLifecycleKeyDistinctness(keys []map[string]any, domain string) error {
	ids, publicKeys, principals := map[string]bool{}, map[string]bool{}, map[string]bool{}
	for _, key := range keys {
		principal := key["principal"].(map[string]any)
		if principal["authority_domain"] != domain {
			return fmt.Errorf("lifecycle key principals must use the root trust domain")
		}
		identity := principalIdentity(principal)
		id, publicKey := key["key_id"].(string), key["public_key_base64url"].(string)
		if ids[id] || publicKeys[publicKey] || principals[identity] {
			return fmt.Errorf("lifecycle key IDs, public keys, and principals must differ")
		}
		ids[id], publicKeys[publicKey], principals[identity] = true, true, true
	}
	return nil
}

func validateIndependentRoots(lifecycle map[string]any, approval *approvalcontract.TrustRoot) error {
	facts, err := approvalcontract.Facts(approval)
	if err != nil {
		return err
	}
	if lifecycle["root_sha256"] == facts.TrustRootSHA256 ||
		lifecycle["trust_domain"] == facts.TrustDomain {
		return fmt.Errorf("approval and lifecycle roots/domains must be independent")
	}
	return compareIndependentKeys(lifecycle["keys"].([]any), facts.RootKeys)
}

func compareIndependentKeys(lifecycle []any, approval []approvalcontract.RootKey) error {
	ids, publicKeys, principals := map[string]bool{}, map[string]bool{}, map[string]bool{}
	for _, key := range approval {
		ids[key.KeyID], publicKeys[key.PublicKeyBase64URL] = true, true
		principals[approvalPrincipalIdentity(key)] = true
	}
	for _, item := range lifecycle {
		key := item.(map[string]any)
		if ids[key["key_id"].(string)] || publicKeys[key["public_key_base64url"].(string)] {
			return fmt.Errorf("approval and lifecycle roots reuse key identity")
		}
		if principals[principalIdentity(key["principal"].(map[string]any))] {
			return fmt.Errorf("approval and lifecycle roots reuse a principal")
		}
	}
	return nil
}

func principalIdentity(node map[string]any) string {
	return node["authority_domain"].(string) + "\x00" + node["principal_id"].(string) +
		"\x00" + node["principal_type"].(string)
}

func approvalPrincipalIdentity(key approvalcontract.RootKey) string {
	return key.AuthorityDomain + "\x00" + key.PrincipalID + "\x00" + key.PrincipalType
}

func lifecycleKey(root map[string]any, usage string) (map[string]any, error) {
	var match map[string]any
	for _, item := range root["keys"].([]any) {
		key := item.(map[string]any)
		if key["usage"] == usage {
			if match != nil {
				return nil, fmt.Errorf("lifecycle root repeats usage %q", usage)
			}
			match = key
		}
	}
	if match == nil {
		return nil, fmt.Errorf("lifecycle root lacks usage %q", usage)
	}
	return match, nil
}

func lifecycleRootKeyView(node map[string]any) RootKey {
	principal := node["principal"].(map[string]any)
	return RootKey{KeyID: node["key_id"].(string), Usage: node["usage"].(string),
		PublicKeyBase64URL: node["public_key_base64url"].(string),
		AuthorityDomain:    principal["authority_domain"].(string),
		PrincipalID:        principal["principal_id"].(string),
		PrincipalType:      principal["principal_type"].(string)}
}

File: harness/authenticated_adr_lifecycle_contract/authority.py
"""Proof-shaped lifecycle authority roots; no signature is verified here."""

from __future__ import annotations

from typing import Any

from authenticated_adr_approval_contract.authority import (
    key_for_usage as approval_key_for_usage,
    validate_signature_profile,
    validate_trust_root as validate_approval_trust_root,
)
from authenticated_adr_approval_contract.shape import (
    fixed_base64url,
    principal_identity,
    validate_principal,
)

from .canonical import ContractError, bounded_canonical_json, self_digest
from .constants import (
    CANONICALIZATION,
    KEY_USAGES,
    MAX_ROOT_BYTES,
    PROFILE_ID,
    REQUEST_KEY_USAGE,
    STATE_KEY_USAGE,
    TRUST_ROOT_API,
    TRUST_ROOT_DOMAIN,
)
from .shape import integer, require_keys, sha256, text

ROOT_FIELDS = {
    "api_version", "canonicalization", "keys", "kind", "profile_id",
    "root_sha256", "signature_profile_sha256", "trust_domain", "trust_epoch",
}
KEY_FIELDS = {"key_id", "principal", "public_key_base64url", "usage"}


def trust_root_sha256(value: dict[str, Any]) -> str:
    return self_digest(TRUST_ROOT_DOMAIN, value, ("root_sha256",),
                       MAX_ROOT_BYTES, "ArchitectureDecisionLifecycleTrustRoot")


def validate_lifecycle_trust_root(value: Any, profile_hash: str) -> dict[str, Any]:
    label = "ArchitectureDecisionLifecycleTrustRoot"
    node = require_keys(value, label, ROOT_FIELDS)
    bounded_canonical_json(node, MAX_ROOT_BYTES, label)
    expected = (TRUST_ROOT_API, CANONICALIZATION, label, PROFILE_ID)
    actual = (node["api_version"], node["canonicalization"], node["kind"],
              node["profile_id"])
    if actual != expected:
        raise ContractError(f"{label} envelope drifted from v1")
    text(node["trust_domain"], f"{label}.trust_domain")
    integer(node["trust_epoch"], f"{label}.trust_epoch", 1)
    if sha256(node["signature_profile_sha256"],
              f"{label}.signature_profile_sha256") != profile_hash:
        raise ContractError(f"{label} does not bind the SignatureProfile")
    _validate_keys(node["keys"], node["trust_domain"])
    if sha256(node["root_sha256"], f"{label}.root_sha256") != trust_root_sha256(node):
        raise ContractError(f"{label} self digest does not match")
    return node


def _validate_keys(value: Any, trust_domain: str) -> None:
    if not isinstance(value, list) or len(value) != 2:
        raise ContractError("lifecycle trust root requires exactly two keys")
    encoded = [bounded_canonical_json(item, MAX_ROOT_BYTES, "lifecycle root key")
               for item in value]
    if encoded != sorted(set(encoded)):
        raise ContractError("lifecycle root keys must be canonical-byte sorted and unique")
    keys = [_validate_key(item, index) for index, item in enumerate(value)]
    if sorted(key["usage"] for key in keys) != sorted(KEY_USAGES):
        raise ContractError("lifecycle trust root requires one key for each exact usage")
    if (len({key["key_id"] for key in keys}) != 2 or
            len({key["public_key_base64url"] for key in keys}) != 2 or
            len({principal_identity(key["principal"]) for key in keys}) != 2):
        raise ContractError("lifecycle key IDs, public keys, and principals must differ")
    if any(key["principal"]["authority_domain"] != trust_domain for key in keys):
        raise ContractError("lifecycle key principals must use the root trust domain")


def validate_independent_roots(lifecycle_root: dict[str, Any],
                               approval_root: dict[str, Any]) -> None:
    if (lifecycle_root["root_sha256"] == approval_root["root_sha256"] or
            lifecycle_root["trust_domain"] == approval_root["trust_domain"]):
        raise ContractError("approval and lifecycle roots/domains must be independent")
    for field in ("key_id", "public_key_base64url"):
        left = {key[field] for key in lifecycle_root["keys"]}
        right = {key[field] for key in approval_root["keys"]}
        if left & right:
            raise ContractError(f"approval and lifecycle roots reuse {field}")
    left_principals = {principal_identity(key["principal"])
                       for key in lifecycle_root["keys"]}
    right_principals = {principal_identity(key["principal"])
                        for key in approval_root["keys"]}
    if left_principals & right_principals:
        raise ContractError("approval and lifecycle roots reuse a principal")


def _validate_key(value: Any, index: int) -> dict[str, Any]:
    label = f"lifecycle_trust_root.keys[{index}]"
    node = require_keys(value, label, KEY_FIELDS)
    text(node["key_id"], f"{label}.key_id")
    if node["usage"] not in KEY_USAGES:
        raise ContractError(f"{label}.usage is unsupported")
    allowed = (("agent", "human", "operator", "service")
               if node["usage"] == REQUEST_KEY_USAGE else ("service",))
    validate_principal(node["principal"], f"{label}.principal", allowed)
    fixed_base64url(node["public_key_base64url"],
                    f"{label}.public_key_base64url", 32)
    return node


def lifecycle_key(root: dict[str, Any], usage: str) -> dict[str, Any]:
    if usage not in KEY_USAGES:
        raise ContractError(f"unsupported lifecycle key usage {usage!r}")
    matches = [key for key in root["keys"] if key["usage"] == usage]
    if len(matches) != 1:
        raise ContractError(f"lifecycle trust root lacks exactly one {usage} key")
    return matches[0]


__all__ = [
    "approval_key_for_usage", "lifecycle_key", "trust_root_sha256",
    "validate_independent_roots",
    "validate_approval_trust_root", "validate_lifecycle_trust_root",
    "validate_signature_profile",
]
